src/libcore/mem/maybe_uninit.rs

```diff
@@ -225,6 +220,41 @@ pub union MaybeUninit<T> {
     value: ManuallyDrop<T>,
 }
 
+/// `Clone` and `Copy` are implementing for `T: Copy` by copying the bytes. But remember that
+/// uninitialized memory is special because the compiler knows that it doesn't have a fixed value.
+/// This example is *incorrect* because we fail to uphold the [initialization invariant][inv]:
+///
+/// ```rust,no_run
+/// # #![allow(invalid_value)]
+/// use std::mem::MaybeUninit;
+///
+/// let x: MaybeUninit<i32> = MaybeUninit::uninit();
+///
+/// // Safe, but the compiler is allowed to omit actually copying anything because the value of x
+/// // is undefined...
```

Comment:
That's a hardware-level view. The spec view is that the compiler *will* definitely copy `x`, but what it copies are the uninitialized bytes, making `x_copy` just as uninitialized as `x`.

But given that `x_copy` is, well, an exact copy of `x`... why is this worth mentioning?^^ There's nothing special going on here. If `x` is 5 then its copy is 5, if `x` is uninitialized then its copy is uninitialized. If anything it would be a big surprise if copying magically initialized things!

So, if you think these docs should be added, then they should make clear that uninit bytes are no an exception here, they just follow the same rules as all other kinds of bytes.

Reply:
Maybe the way to teach this would be a section at the top of the docs along the lines of "What is uninitialized memory?" that explained this "true/false/uninit" paradigm. It could mention how this is a specification level thing, that, for instance, Miri does in fact implement directly. 

My intent with the copying example was to make clear how this specification level "tri-state" results in miscompilation, and how it's *not* as simple as "the compiler lets the bits be uninit"

Reply:
I agree that properly explaining what you want to explain here requires some kind of introduction to "explicit uninitialized bytes". You talk about tri-state, so it seems like you are thinking of a bit-level property. I am not sure which is easier to teach, but Miri implements a byte-level notion of "uninitialized", and also bytes are generally the "smallest possible piece of data" that a Rust program can manipulate, so I think bytes work better than bits here. I am not sure though.

However, what would be the best place to teach this? Putting this into the API docs of a libcore type seems somewhat odd. Cc @rust-lang/wg-unsafe-code-guidelines @Centril

Reply:
Indeed, it looks like rustc itself works at roughly the structure level: initializing a single byte in a structure "freezes" the rest of the bytes. But you certainly wouldn't want to rely on that.

A big problem with talking about bits is people might get the wrong idea with bitflags, and try to set a single bit's state in an uninitialized byte; in Rust that'll be a read-modify-write, reading uninit bytes.

Maybe describe it as "At the specification level, a Rust byte is *either* an initialized 8 bit integer, *or* the special state "uninitialized""

Reply:
> However, what would be the best place to teach this? Putting this into the API docs of a libcore type seems somewhat odd. Cc @rust-lang/wg-unsafe-code-guidelines @Centril

Somewhat odd yes, yet I find myself in favor because the subject is not well understood in general and because it is a good place for people to be introduced to it. Maybe you could extract the most salient bits from your blog post and link to it?

Reply:
@Centril Whose blog post?

Reply:
I think @Centril means [this post of mine](https://www.ralfj.de/blog/2019/07/14/uninit.html).

Reply:
> Maybe describe it as "At the specification level, a Rust byte is either an initialized 8 bit integer, or the special state "uninitialized""

Yeah, something like that.
To be on the safe side we should probably have a footnote saying that reality is actually even more complicated and other exotic forms of Rust bytes exist (as described in [this blog post](https://www.ralfj.de/blog/2018/07/24/pointers-and-bytes.html)) -- just to make sure that people don't hold us to this later.

Reply:
I think we should put some discussion here and then link to a section where it is described in more detail, although that might be a placeholder issue for the time being (and the issue can link to Ralf's blog post)